docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,814 @@
+}
+```
+
+Given the above definitions, a catalog can finally be created as follow:
+
+```go
+var c = catalog.NewCatalog()
+
+func init() {
+    v1alpha1.AddToCatalog(c)
+    v1alpha2.AddToCatalog(c)
+    v1alpha3.AddToCatalog(c)
+}
+```
+
+The catalog provides the core knowledge required to manage all the Runtime Hooks supported by Cluster API;
+the first application of such knowledge will be to retrieve all the info required to generate the OpenAPI specification
+for Runtime Hooks with a dedicated tool under `hack/tools`.
+
+### Discovering Runtime Extensions
+
+_Note: the controller described in this paragraph will be executed only if the Runtime Extension feature flag is set to true._
+
+Cluster API is going to implement a new controller that looks at Runtime Extension Configurations; the main
+responsibility of this controller should be to maintain an internal, shared **registry** of available extensions
+at a given time.
+
+Please note that the Runtime Extensions registry also provides a single point to centralize a set of common behaviors
+supporting interaction with those external components, thus making the adoption of this feature scalable - 
+in the sense of being used for an increasing numbers of use cases in Cluster API - while operating consistently
+across the board.
+
+A first behavior that falls into this category is the implementation of exponential backoff mechanisms
+in case of errors, thus preventing Cluster API from creating pressure on HTTP Servers recovering from or with
+ongoing operational issues.
+
+Another cross-cutting concern is about ensuring that Runtime Extensions, which are external components triggered
+in the middle of Cluster API controllers logic, do not block the reconciliation process indefinitely 
+(e.g by enforcing a maximum timeout for all the Runtime Extensions calls).
+
+### Calling Runtime Extensions
+
+_Note: the code described in this paragraph will be executed only if the Runtime Extension feature flag is set to true._
+
+Cluster API is going to implement calls to registered Runtime Extensions at well-known moments of the Cluster’s lifecycle.
+
+The two key elements that make the implementation of runtime extension calls simple and consistent across
+the codebase are:
+
+- The Runtime Hook catalog, providing the info about all the defined Runtime Hooks, supported version and
+  corresponding input/output types;
+- The Runtime Extensions registry, providing info about the registered Runtime Extensions implementing the
+  Runtime Hooks defined above.
+
+Given these two elements, the code for calling Runtime Extensions is:
+
+_Note: this is pseudo code, meant to demonstrate a set of elements described below the example; the exact details
+will be defined during implementation, possibly taking advantage of golang generics._
+
+```go
+extensions := registry.Get(
+    registry.Group("cluster.runtime.cluster.x-k8s.io/v1alpha2”),
+    registry.Hook(&v1alpha2.BeforeUgradeHook{}),
```

Comment:
Instead of using the "hub" version of the hook as the identifier should we consider using something else instead. I am asking to 

Consider a case where a hook is dropped in a new version but is still present in the old versions  - in such a case using the last version as the "hub" version to identify the list of hooks will not be ideal. Should we explore some other way to maintain a list of all available hook (across all versions). The version specific definitions of the hooks will then be registered against this new identifier.

Reply:
As written in the note this is pseudo code, so details are up to changes but the TL;DR; is that we should implement the call operation only once vs "duplicating" the code for each new version.
Said that, a possible way forward for the use case when a hook is dropped is to consider the last available version of that hook as "Hub" (there is no single "Hub" version like in APIs, you can choose your preferred version for each hook); other solutions can be figured out during implementation.

Reply:
+1 We should dive into those details during implementation. But I don't think it's a big issue. 

For now it seems to be good enough to just define the latest existing version as the hub for conversion.